File: food-app-backend-js/src/controllers/order.controller.js
import orderService from '../services/order.service.js';

class OrderController {
  async createOrderWithPayment(req, res) {
    try {
      const response = await orderService.createOrderWithPayment(req.body, req.userId);
      return res.status(201).json(response);
    } catch (error) {
      console.error('Create Order Error:', error);
      return res.status(500).json({ message: error.message || 'Failed to create order.' });
    }
  }

  async verifyPayment(req, res) {
    const verificationData = req.body || {};
    const paymentIntentId = verificationData.stripePaymentIntentId;
    const status = verificationData.status || 'UNKNOWN';

    if (!paymentIntentId) {
      return res.status(400).json({ message: 'stripePaymentIntentId is required.' });
    }

    try {
      await orderService.verifyPayment(verificationData, status);
      return res.status(200).json({ message: `Payment verified successfully for PI: ${paymentIntentId}` });
    } catch (error) {
      console.error('Verify Payment Error:', error);
      return res.status(400).json({ message: error.message || 'Payment verification failed.' });
    }
  }

  async getUserOrder(req, res) {
    try {
      const response = await orderService.getUserOrders(req.userId);
      return res.status(200).json(response);
    } catch (error) {
      console.error('Get User Orders Error:', error);
      return res.status(500).json({ message: 'Failed to retrieve user orders.' });
    }
  }

  async removeOrder(req, res) {
    try {
      await orderService.removeOrder(req.params.orderId);
      return res.status(204).end();
    } catch (error) {
      console.error('Remove Order Error:', error);
      return res.status(404).json({ message: error.message || 'Order not found or deletion failed.' });
    }
  }

  async getOrdersOfAllUsers(req, res) {
    try {
      const response = await orderService.getOrdersOfAllUsers();
      return res.status(200).json(response);
    } catch (error) {
      console.error('Get All Orders Error:', error);
      return res.status(500).json({ message: 'Failed to retrieve all orders.' });
    }
  }

  async updateOrder(req, res) {
    const status = req.body.orderStatus;
    if (!status) return res.status(400).json({ message: 'Order status is required.' });

    try {
      const response = await orderService.updateOrder(req.params.orderId, status);
      return res.status(200).json(response);
    } catch (error) {
      console.error('Update Order Error:', error);
      if ((error.message || '').includes('Invalid order status')) {
        return res.status(400).json({ message: error.message });
      }
      return res.status(404).json({ message: error.message || 'Order not found.' });
    }
  }
}

export default new OrderController();

File: food-app-backend-java/src/main/java/com/kspk/controller/OrderController.java
package com.kspk.controller;

import com.kspk.DTOs.OrderRequest;
import com.kspk.DTOs.OrderResponse;
import com.kspk.service.OrderService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/orders")
public class OrderController {

    @Autowired
    private OrderService orderService;

    @PostMapping("/create")
    public ResponseEntity<OrderResponse> createOrderWithPayment(@RequestBody OrderRequest request) {
        return ResponseEntity.ok(orderService.createOrderWithPayment(request));
    }

    @PostMapping("/verify")
    public ResponseEntity<String> verifyPayment(@RequestBody Map<String, String> verificationData) {
        String paymentIntentId = verificationData.get("stripePaymentIntentId");
        String status = verificationData.getOrDefault("status", "UNKNOWN");
        orderService.varifyPayment(verificationData, status);
        return ResponseEntity.ok("✅ Payment verified successfully for PI: " + paymentIntentId);
    }

    @GetMapping
    public ResponseEntity<List<OrderResponse>> getUserOrder() {
        return ResponseEntity.ok(orderService.getUserOrders());
    }

    @DeleteMapping("/{orderId}")
    public ResponseEntity<Void> removeOrder(@PathVariable String orderId) {
        orderService.removeOrder(orderId);
        return ResponseEntity.ok().build();
    }

    // ✅ Admin-only endpoints
    @PreAuthorize("hasRole('ADMIN')")
    @GetMapping("/all")
    public ResponseEntity<List<OrderResponse>> getOrdersOfAllUsers() {
        return ResponseEntity.ok(orderService.getOrdersOfAllUsers());
    }

    @PreAuthorize("hasRole('ADMIN')")
    @PutMapping("/{orderId}")
    public ResponseEntity<OrderResponse> updateOrder(
            @PathVariable String orderId,
            @RequestBody Map<String, String> request
    ) {
        String status = request.get("orderStatus");
        return ResponseEntity.ok(orderService.updateOrder(orderId, status));
    }

}
